packages/firebase_database/lib/ui/firebase_list.dart

```diff
@@ -85,10 +85,18 @@ class FirebaseList extends ListBase<DataSnapshot>
 
   int _indexForKey(String key) {
     assert(key != null);
-    // TODO(jackson): We could binary search since the list is already sorted.
-    for (int index = 0; index < _snapshots.length; index++) {
-      if (key == _snapshots[index].key) {
-        return index;
+    int min = 0;
+    int max = _snapshots.length;
+    int mid;
+    while (min < max) {
+      mid = min + ((max - min) >> 1);
+      if (key == _snapshots[mid].key) {
+        return mid;
+      }
+      if (key.compareTo(_snapshots[mid].key) > 0) {
+        min = mid;
```

Comment:
The invariant of the `while` loop is that the wanted key, if present, is in the half-open interval [min..max[

Then we need to update `min` to `mid + 1` in line 97.

I think we get an infinite loop, if `mid == min`. Which happens if `max == min + 1` and the key is not present.

Writing binary search correctly is tricky enough to warrant a unit test.

Reply:
Thanks, @mravn-google. It requires more understanding of Flutter unit test case for me (newbie in Flutter). I will come back on this.

Reply:
@manishkumar-dsi Let us know if you would prefer us taking it from here.

Reply:
@manishkumar-dsi A rudimentary unit test is work in progress: https://github.com/flutter/plugins/pull/459

Reply:
Sure @mravn-google. Please go ahead. Thanks